tests/suites/test_suite_pkwrite.function

```diff
@@ -2,88 +2,90 @@
-/* END_HEADER */
 
-/* BEGIN_DEPENDENCIES
- * depends_on:MBEDTLS_PK_PARSE_C:MBEDTLS_PK_WRITE_C:MBEDTLS_BIGNUM_C:MBEDTLS_FS_IO
- * END_DEPENDENCIES
- */
+typedef enum {
+    TEST_PEM,
+    TEST_DER
+} pkwrite_file_format_t;
 
-/* BEGIN_CASE depends_on:MBEDTLS_PEM_WRITE_C */
-void pk_write_pubkey_check(char *key_file)
+static void pk_write_check_common(char *key_file, int is_public_key, int is_der)
 {
     mbedtls_pk_context key;
-    unsigned char buf[5000];
-    unsigned char check_buf[5000];
+    unsigned char *buf = NULL;
+    unsigned char *check_buf = NULL;
+    unsigned char *start_buf;
+    size_t buf_len, check_buf_len;
     int ret;
-    FILE *f;
-    size_t ilen, pem_len, buf_index;
 
-    memset(buf, 0, sizeof(buf));
-    memset(check_buf, 0, sizeof(check_buf));
+    /* Note: if mbedtls_pk_load_file() successfully reads the file, then
+       it also allocates check_buf, which should be freed on exit */
+    TEST_EQUAL(mbedtls_pk_load_file(key_file, &check_buf, &check_buf_len), 0);
+    TEST_ASSERT(check_buf_len > 0);
 
-    mbedtls_pk_init(&key);
-    TEST_ASSERT(mbedtls_pk_parse_public_keyfile(&key, key_file) == 0);
-
-    ret = mbedtls_pk_write_pubkey_pem(&key, buf, sizeof(buf));
-    TEST_ASSERT(ret == 0);
+    ASSERT_ALLOC(buf, check_buf_len);
 
-    pem_len = strlen((char *) buf);
-
-    // check that the rest of the buffer remains clear
-    for (buf_index = pem_len; buf_index < sizeof(buf); ++buf_index) {
-        TEST_ASSERT(buf[buf_index] == 0);
+    mbedtls_pk_init(&key);
+    if (is_public_key) {
+        TEST_EQUAL(mbedtls_pk_parse_public_keyfile(&key, key_file), 0);
+        if (is_der) {
+            ret = mbedtls_pk_write_pubkey_der(&key, buf, check_buf_len);
+        } else {
+#if defined(MBEDTLS_PEM_WRITE_C)
+            ret = mbedtls_pk_write_pubkey_pem(&key, buf, check_buf_len);
+#else
+            ret = MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE;
+#endif
+        }
+    } else {
+        TEST_EQUAL(mbedtls_pk_parse_keyfile(&key, key_file, NULL,
+                                            mbedtls_test_rnd_std_rand, NULL), 0);
+        if (is_der) {
+            ret = mbedtls_pk_write_key_der(&key, buf, check_buf_len);
+        } else {
+#if defined(MBEDTLS_PEM_WRITE_C)
+            ret = mbedtls_pk_write_key_pem(&key, buf, check_buf_len);
+#else
+            ret = MBEDTLS_ERR_PK_FEATURE_UNAVAILABLE;
+#endif
+        }
     }
 
-    f = fopen(key_file, "r");
-    TEST_ASSERT(f != NULL);
-    ilen = fread(check_buf, 1, sizeof(check_buf), f);
-    fclose(f);
+    if (is_der) {
+        TEST_LE_U(1, ret);
+        buf_len = ret;
+        start_buf = buf + check_buf_len - buf_len;
+    } else {
+        TEST_EQUAL(ret, 0);
+        buf_len = strlen((char *) buf) + 1; /* +1 takes the string terminator into account */
+        start_buf = buf;
+    }
 
-    TEST_ASSERT(ilen == pem_len);
-    TEST_ASSERT(memcmp((char *) buf, (char *) check_buf, ilen) == 0);
+    ASSERT_COMPARE(start_buf, buf_len, check_buf, check_buf_len);
```

Comment:
As Manuel suspects, this fails with PEM files on Windows because of its handling of text files. `mbedtls_pk_write_key_pem` writes to the buffer, without going through the I/O layer, so newlines in `start_buf` are a `\n` character. `mbedtls_pk_load_file` loads files in binary mode (it has no choice since it must support DER), so newlines in `check_buf` are `\n` on non-Windows but `\r\n` on Windows (because Git treats PEM files as text, so it converts `\n` to `\r\n` upon checkout).

This wasn't a problem with the old code because it called `fopen` in text mode to read `check_buf`, so the I/O layer converted `\r\n` in the file to `\n` in memory. But of course that would have corrupted DER files.

I guess the solution is to remove CR characters from `check_buf` if the format is PEM. Or keep calling `fopen` directly, but choose text vs binary mode depending on whether the data is DER or PEM.

Reply:
Thanks a lot for your help! It really saved me a lot of time since I was not aware of this difference and I didn't have a Windows machine at hand.

Reply:
We got hit by this on another PR not so long ago. While I understand why git it doing it, I still find it always surprising that what we check out (on Windows) is not byte-for-byte identical to what we commited (on Linux), at least with the default settings. Thanks Gilles for the reminder!